### PR TITLE
New Authorization Behavior - Use current task rather than new task

### PR DESCRIPTION
**Pull request by shoatman**
- Added a new CurrentTaskBrowserTabActivity to receive the result of the authorization request
- Modified the code that looks for other apps having registered the redirect URI intent filer to account for this new activity
- configured Msalconfiguration singleton in PCAConfig initializeInternal method.

### CODE REVIEW

File: msal/src/main/java/com/microsoft/identity/client/CurrentTaskBrowserTabActivity.java [hunk 1]
```diff
@@ -0,0 +1,118 @@
...
+ *     &lt;category android:name="android.intent.category.DEFAULT" /&gt;
+ *
+ *     The target activity allows itself to be started by a web browser to display data.
+ *     &lt;category android:name="android.intent.category.BROWSABLE" /&gt;
+ *
+ *     BrowserTabActivity will be launched when matching the custom url scheme.
+ *     &lt;data android:scheme="msalclientid" android:host="auth" /&gt;
+ * &lt;/intent-filter&gt;
+ * </pre>
+ */
```
Comment (shahzaibj) [inline msal/src/main/java/com/microsoft/identity/client/CurrentTaskBrowserTabActivity.java:65]: The formatting looks strange

File: msal/src/main/java/com/microsoft/identity/client/CurrentTaskBrowserTabActivity.java [hunk 2]
```diff
@@ -0,0 +1,118 @@
...
+public final class CurrentTaskBrowserTabActivity extends Activity {
+    private static final int REDIRECT_RECEIVED_CODE = 2;
+    private BroadcastReceiver mCloseBroadcastReceiver;
+
+
```
Comment (shahzaibj) [inline msal/src/main/java/com/microsoft/identity/client/CurrentTaskBrowserTabActivity.java:70]:
```suggestion

```

File: msal/src/main/AndroidManifest.xml [hunk 3]
```diff
@@ -13,16 +13,26 @@
             android:exported="false"
             android:launchMode="singleTask" />
 
+        <activity
```
Comment (iamgusain) [inline msal/src/main/AndroidManifest.xml:16]: ><activity [](http://example.com/codeflow?start=8&length=9)

Can we add a description of this, as is done for other activities

File: msal/src/main/java/com/microsoft/identity/client/CurrentTaskBrowserTabActivity.java [hunk 4]
```diff
@@ -0,0 +1,118 @@
...
+import static com.microsoft.identity.common.adal.internal.AuthenticationConstants.AuthorizationIntentAction.DESTROY_REDIRECT_RECEIVING_ACTIVITY;
+
+
+/**
+ * MSAL activity class (needs to be public in order to be discoverable by the os) to get the browser redirect with auth code from authorize
+ * endpoint. This activity has to be exposed by "android:exported=true", and intent filter has to be declared in the
+ * manifest for the activity.
+ * <p>
+ * When the AuthorizationAgent is launched, and we're redirected back with the redirect
+ * uri (the redirect must be unique across apps on a device), the os will fire an intent with the redirect,
```
Comment (iamgusain) [inline msal/src/main/java/com/microsoft/identity/client/CurrentTaskBrowserTabActivity.java:49]: >redirect [](http://example.com/codeflow?start=98&length=8)

nit: redirect uri?

**Comment by iamgusain** [comment]
>                         String.format("Another application %s is listening for the URL scheme %s", activityInfo.packageName, url)

Should we also log "activityInfo.name" here.

---
Refers to: msal/src/main/java/com/microsoft/identity/client/PublicClientApplicationConfiguration.java:632 in 5412023. [](commit_id = 54120237686b3b55c60ad403cc2585e9b7cb4fef, deletion_comment = False)

File: msal/src/main/AndroidManifest.xml (hunk 3 above)
Comment (iamgusain) [reply msal/src/main/AndroidManifest.xml:16]: oh i see.. its the same as above.. nvm

File: msal/src/main/java/com/microsoft/identity/client/CurrentTaskBrowserTabActivity.java [hunk 5]
```diff
@@ -0,0 +1,118 @@
...
+ * and the BrowserTabActivity will be launched.
```
Comment (iamgusain) [inline msal/src/main/java/com/microsoft/identity/client/CurrentTaskBrowserTabActivity.java:50]: >BrowserTabActivity will be launched [](http://example.com/codeflow?start=11&length=35)

Curious to know how it works.
Will it launch the CurrentTaskBrowserTabActivity or the BrowserTabActivity as both will be listening for the same redirect uri, right?

File: msal/src/main/java/com/microsoft/identity/client/CurrentTaskBrowserTabActivity.java [hunk 6]
```diff
@@ -115,4 +135,14 @@ protected void onDestroy() {
         LocalBroadcastManager.getInstance(this).unregisterReceiver(mCloseBroadcastReceiver);
         super.onDestroy();
     }
+
+    private String decodeState(final String encodedState) {
```
Comment (iambmelt) [inline msal/src/main/java/com/microsoft/identity/client/CurrentTaskBrowserTabActivity.java:139]:
```suggestion
    @Nullable
    private String decodeState(final String encodedState) {
```

File: msal/build.gradle [hunk 7]
```diff
@@ -291,12 +291,14 @@ dependencies {
         transitive = true
     }
 
+    /*
```
Comment (iambmelt) [inline msal/build.gradle:294]: Something we can jettison?

File: msal/src/main/AndroidManifest.xml [hunk 8]
```diff
@@ -13,16 +13,26 @@
...
         <!-- MSAL will use system webview (custom tab) to render the sign-in page, BrowserTabActivity is to get response back from System Webview. Multiple apps on the device could integrate MSAL, OS will check which BrowserTabActivity can handle the intent based on the intent filter declared, so this activity has to be exported.-->
         <activity
             android:name="com.microsoft.identity.client.BrowserTabActivity"
             android:exported="true" />
+
+        <activity
+            android:name="com.microsoft.identity.client.CurrentTaskBrowserTabActivity"
+            android:exported="true" />
     </application>
 </manifest>
```
Comment (iambmelt) [inline msal/src/main/AndroidManifest.xml:38]:
```suggestion
</manifest>

```

Adding terminating newline

File: msal/src/main/java/com/microsoft/identity/client/CurrentTaskBrowserTabActivity.java [hunk 9]
```diff
@@ -0,0 +1,148 @@
...
+        if (StringUtil.isEmpty(encodedState)) {
+            Logger.warn(TAG, "Decode state failed because the input state is empty.");
+            return null;
+        }
+
+        final byte[] stateBytes = Base64.decode(encodedState, Base64.NO_PADDING | Base64.URL_SAFE);
+        return new String(stateBytes, Charset.defaultCharset());
+    }
+}
```
Comment (iambmelt) [inline msal/src/main/java/com/microsoft/identity/client/CurrentTaskBrowserTabActivity.java:148]:
```suggestion
}

```

File: testapps/testapp/src/main/res/raw/msal_config_default.json [hunk 10]
```diff
@@ -2,8 +2,9 @@
   "client_id" : "4b0db8c2-9f26-4417-8bde-3f0e3656f8e0",
   "authorization_user_agent" : "DEFAULT",
   "redirect_uri" : "msauth://com.msft.identity.client.sample.local/1wIqXSqBj7w%2Bh11ZifsnqwgyKrY%3D",
-  "handle_null_taskaffinity": true,
+  "handle_null_taskaffinity": false,
```
Comment (iambmelt) [inline testapps/testapp/src/main/res/raw/msal_config_default.json:5]: Do we have new docs covering this?

File: msal/src/main/AndroidManifest.xml [hunk 11]
```diff
@@ -13,16 +13,26 @@
...
             android:configChanges="orientation|keyboardHidden|screenSize|smallestScreenSize|screenLayout"
             android:exported="false"
-            android:launchMode="singleTask" />
+            android:launchMode="standard" />
 
         <!-- MSAL will use system webview (custom tab) to render the sign-in page, BrowserTabActivity is to get response back from System Webview. Multiple apps on the device could integrate MSAL, OS will check which BrowserTabActivity can handle the intent based on the intent filter declared, so this activity has to be exported.-->
         <activity
             android:name="com.microsoft.identity.client.BrowserTabActivity"
             android:exported="true" />
+
```
Comment (rpdome) [inline msal/src/main/AndroidManifest.xml:33]: We might want to call out explicitly somewhere what are the different between these two (BrowserTabActivity vs CurrentTaskBrowserTabActivity)

File: msal/src/main/java/com/microsoft/identity/client/CurrentTaskBrowserTabActivity.java [hunk 12]
```diff
@@ -0,0 +1,148 @@
...
+
+import java.nio.charset.Charset;
+import java.util.HashMap;
+
+import static com.microsoft.identity.common.adal.internal.AuthenticationConstants.AuthorizationIntentAction.DESTROY_REDIRECT_RECEIVING_ACTIVITY;
+import static com.microsoft.identity.common.adal.internal.AuthenticationConstants.AuthorizationIntentAction.REDIRECT_RETURNED_ACTION;
+
+
+/**
+ * MSAL activity class (needs to be public in order to be discoverable by the os) to get the browser redirect with auth code from authorize
```
Comment (rpdome) [inline msal/src/main/java/com/microsoft/identity/client/CurrentTaskBrowserTabActivity.java:49]: Is the comment updated?

File: msal/src/main/java/com/microsoft/identity/client/CurrentTaskBrowserTabActivity.java [hunk 13]
```diff
@@ -0,0 +1,148 @@
...
+                com.microsoft.identity.common.logging.Logger.error(TAG, "Did not receive task id in the authorization response", ex);
+                throw ex;
+            }
+        }
+
+        if (savedInstanceState == null
+                && getIntent() != null
+                && !StringUtil.isEmpty(getIntent().getDataString())) {
+            //Leaving this for now to set static response URI value
+            final Intent responseIntent = CurrentTaskBrowserAuthorizationFragment.createCustomTabResponseIntent(this, response);
```
Comment (rpdome) [inline msal/src/main/java/com/microsoft/identity/client/CurrentTaskBrowserTabActivity.java:98]: I assume this has a dependency on another common PR which is still in draft?

File: testapps/testapp/src/main/res/raw/msal_config_default.json [hunk 14]
```diff
@@ -2,8 +2,9 @@
...
   "concurrent_authorization_requests": true,
+  "authorization_in_current_task": true,
```
Comment (rpdome) [inline testapps/testapp/src/main/res/raw/msal_config_default.json:7]: Do we want this to be = true by default?

File: testapps/testapp/src/main/res/raw/msal_config_default.json (hunk 10 above)
Comment (shoatman) [reply testapps/testapp/src/main/res/raw/msal_config_default.json:5]: I have to write these docs specifically for Samsung... not sure that we want to document this in public yet.

File: msal/src/main/AndroidManifest.xml (hunk 11 above)
Comment (shoatman) [reply msal/src/main/AndroidManifest.xml:33]: They have a different implementation of the activity... in terms of the behaviors they perform.  I'll write up some internal documentation for this as well as some external documentation for consumers.

File: msal/src/main/java/com/microsoft/identity/client/CurrentTaskBrowserTabActivity.java (hunk 13 above)
Comment (shoatman) [reply msal/src/main/java/com/microsoft/identity/client/CurrentTaskBrowserTabActivity.java:98]: Yes.  This PR isn't into dev.

File: testapps/testapp/src/main/res/raw/msal_config_default.json (hunk 14 above)
Comment (shoatman) [reply testapps/testapp/src/main/res/raw/msal_config_default.json:7]: Probably not for the test app.  I need to add an automation run for this to our test plan/pipeline.

**Pull request closed without merging**